invenio_drafts_resources/records/api.py

```diff
@@ -33,57 +55,66 @@ class ParentRecord(RecordBase):
     model_cls = None
-
+    #: The parent state's SQLAlchemy model class. Must be set by the subclass.
+    versions_model_cls = None
+    #: The parent record's API class. Must be set by the subclass.
+    parent_record_cls = None
+
+    #
+    # System fields
+    #
+    #: The internal persistent identifier. Records and drafts share UUID.
     pid = PIDField('id', provider=DraftRecordIdProviderV2, delete=True)
 
+    #: System field to check if a record has been published.
     is_published = PIDStatusCheckField(status=PIDStatus.REGISTERED)
 
-    #: List of field names (strings) to copy from the draft on create.
-    create_from_draft = []
+    #: The parent record - the draft is responsible for creating the parent.
+    parent = ParentField(
+        ParentRecord, create=False, soft_delete=False, hard_delete=False)
 
-    #: List of field names (strings) to copy from the draft on update.
-    update_from_draft = []
+    #: Version relationship
+    versions = VersionsField(create=True, set_latest=True)
 
-    # TODO: Below three methods create_from(), update_from() and register() has
-    # to be refactored. They are accessing "parent" but it may not be defined.
-    # Instead, this work should be delegated to the system fields, but it's not
-    # easy to add a new pre/post_create_from pre/post_register hook.
     @classmethod
-    def create_from(cls, draft):
-        """Create a new record from a draft."""
-        record = cls.create(
-            {},
-            id_=draft.id,
-            pid=draft.pid,
-            **{f: getattr(draft, f) for f in cls.create_from_draft}
-        )
-
-        # NOTE: Merge pid into the current db session if not already in the
-        # session.
-        cls.pid.session_merge(record)
-        record.parent.__class__.pid.session_merge(record.parent)
-
-        record.update_from(draft)
-
+    def publish(cls, draft):
+        """Publish a draft as a new record.
+
+        If a record already exists, we simply get the record. If a draft has
+        not yet been published, we create the record.
+
+        The caller is responsible for registering the internal persistent
+        identifiers (see ``register()``).
+        """
+        if draft.is_published:
+            record = cls.get_record(draft.id)
+        else:
+            record = cls.create(
+                {},
+                # A draft and record share UUID, so we reuse the draft's id/pid
+                id_=draft.id,
+                pid=draft.pid,
+                # Link the record with the parent record and set the versioning
+                # relationship.
+                parent=draft.parent,
+                versions=draft.versions,
+            )
+            # Merge the PIDs into the current db session if not already in the
+            # session.
+            cls.pid.session_merge(record)
+            cls.parent_record_cls.pid.session_merge(record.parent)
         return record
 
-    def update_from(self, draft):
```

Comment:
Comment: All updates of data is now the responsibility of a component (see e.g. MetadataComponent). Doing record.update(**draft) is very unreliable and bypasses all the system fields.